napi-inl.h

```diff
@@ -3618,6 +3618,275 @@ inline void AsyncWorker::OnWorkComplete(
+                                                  Finalizer finalizeCallback,
+                                                  Context* context) {
+  static_assert(details::can_make_string<ResourceString>::value
+      || std::is_convertible<ResourceString, napi_value>::value,
+      "Resource name should be string convertible type");
+
+  napi_threadsafe_function tsFunctionValue;
+  auto* finalizeData = new details::FinalizeData<DataType, Finalizer>({
+      finalizeCallback, context });
+  napi_status status = napi_create_threadsafe_function(env, callback, resource,
+      Value::From(env, resourceName), maxQueueSize, initialThreadCount, data,
+      details::FinalizeData<DataType, Finalizer, Context>::WrapperWithHint,
+      finalizeData, CallJS, &tsFunctionValue);
+  if (status != napi_ok) {
+    delete finalizeData;
+    NAPI_THROW_IF_FAILED(env, status, ThreadSafeFunction());
+  }
+
+  return ThreadSafeFunction(env, tsFunctionValue);
+}
+
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall() const {
+  return CallInternal(nullptr, napi_tsfn_blocking);
+}
+
+template <typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall(
+    Callback callback) const {
+  return CallInternal(new CallbackWrapper(callback), napi_tsfn_blocking);
+}
+
+template <typename DataType, typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall(
+    DataType* data, Callback callback) const {
+  auto wrapper = [data, callback](Env env, Function jsCallback) {
+    callback(env, jsCallback, data);
+  };
+  return CallInternal(new CallbackWrapper(wrapper), napi_tsfn_blocking);
+}
+
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall() const {
+  return CallInternal(nullptr, napi_tsfn_nonblocking);
+}
+
+template <typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall(
+    Callback callback) const {
+  return CallInternal(new CallbackWrapper(callback), napi_tsfn_nonblocking);
+}
+
+template <typename DataType, typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall(
+    DataType* data, Callback callback) const {
+  auto wrapper = [data, callback](Env env, Function jsCallback) {
+    callback(env, jsCallback, data);
+  };
+  return CallInternal(new CallbackWrapper(wrapper), napi_tsfn_nonblocking);
+}
+
+inline bool ThreadSafeFunction::Acquire() const {
+  return !IsAborted() && napi_acquire_threadsafe_function(
+      _tsFunctionValue) == napi_ok;
+}
+
+inline bool ThreadSafeFunction::Release() {
+  return !IsAborted() && napi_release_threadsafe_function(
+      _tsFunctionValue, napi_tsfn_release) == napi_ok;
+}
+
+inline bool ThreadSafeFunction::Abort() {
+  if (IsAborted()) {
+    return false;
+  }
+
+  napi_status status = napi_release_threadsafe_function(
+      _tsFunctionValue, napi_tsfn_abort);
+
+  _tsFunctionValue = nullptr;
+  _env = nullptr;
+
+  return status == napi_ok;
```

Comment:
This should specifically handle `napi_ok`, `napi_generic_failure`, and `napi_invalid_arg`.

Reply:
Sorry, I don't understand what you want exactly. Do you want this function to return three status instead of boolean? (like napi_release_threadsafe_function) If so, could you please explain the use cases in more details?

Reply:
@gabrielschulhof Gentle ping

Reply:
@romandev sorry about the delay! Normally, on the main thread, we'd throw an exception if the status is anything other than `napi_ok`. Since this may or may not be the main thread, I think we should pass the `napi_status` through. In fact, we should probably not check `IsAborted()`, because that duplicates what `napi_threadsafe_function` already does. `napi_threadsafe_function` will return `napi_invalid_arg` if there are no more threads associated with it. Additionally, `napi_release_threadsafe_function` internally initiates a `uv_async_send` which may result in a `napi_generic_failure`. This status should not be swallowed.

I guess what I'm saying is that I don't believe that in this case we should do anything more than pass through what the underlying C API produces, because we do not have the facilities to throw errors on behalf of the user like we do on the main thread.